clippy_lints/src/manual_rem_euclid.rs

```diff
@@ -0,0 +1,128 @@
+    /// Use instead:
+    /// ```rust
+    /// let x: i32 = 24;
+    /// let rem = x.rem_euclid(4);
+    /// ```
+    #[clippy::version = "1.63.0"]
+    pub MANUAL_REM_EUCLID,
+    complexity,
+    "manually reimplementing `rem_euclid`"
+}
+
+pub struct ManualRemEuclid {
+    msrv: Option<RustcVersion>,
+}
+
+impl ManualRemEuclid {
+    #[must_use]
+    pub fn new(msrv: Option<RustcVersion>) -> Self {
+        Self { msrv }
+    }
+}
+
+impl_lint_pass!(ManualRemEuclid => [MANUAL_REM_EUCLID]);
+
+impl<'tcx> LateLintPass<'tcx> for ManualRemEuclid {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) {
+        if !meets_msrv(self.msrv, msrvs::REM_EUCLID) {
+            return;
+        }
+
+        if in_constant(cx, expr.hir_id) && !meets_msrv(self.msrv, msrvs::REM_EUCLID_CONST) {
+            return;
+        }
+
+        if let ExprKind::Binary(op1, ..) = expr.kind
+            && op1.node == BinOpKind::Rem
+            && let Some((const1, expr1)) = check_for_positive_int_constant(cx, expr, false)
+            && let ExprKind::Binary(op2, ..) = expr1.kind
+            && op2.node == BinOpKind::Add
+            && let Some((const2, expr2)) = check_for_positive_int_constant(cx, expr1, true)
+            && let ExprKind::Binary(op3, ..) = expr2.kind
+            && op3.node == BinOpKind::Rem
+            && let Some((const3, expr3)) = check_for_positive_int_constant(cx, expr2, false)
+            && const1 == const2 && const2 == const3
+            && let Some(hir_id) = path_to_local(expr3)
+            && let Some(Node::Binding(_)) = cx.tcx.hir().find(hir_id) {
+                // Apply only to params or locals with annotated types
+                match cx.tcx.hir().find(cx.tcx.hir().get_parent_node(hir_id)) {
+                    Some(Node::Param(..)) => (),
+                    Some(Node::Local(local)) => {
+                        let Some(ty) = local.ty else { return };
+                        if matches!(ty.kind, TyKind::Infer) {
+                            return;
+                        }
+                    }
+                    _ => return,
+                };
+
+                let mut app = Applicability::MachineApplicable;
+                let rem_of = snippet_with_applicability(cx, expr3.span, "_", &mut app);
+                span_lint_and_sugg(
+                    cx,
+                    MANUAL_REM_EUCLID,
+                    expr.span,
+                    "manual `rem_euclid` implementation",
+                    "consider using",
+                    format!("{rem_of}.rem_euclid({const1})"),
+                    app,
+                );
+        }
+    }
+
+    extract_msrv_attr!(LateContext);
+}
+
+// Takes a binary expression and separates the operands into the int constant and the other
+// operand. Ensures the int constant is positive. Operators that are not commutative must have the
+// constant appear on the right hand side to return a value.
+fn check_for_positive_int_constant<'a>(
```

Comment:
This would be better split into two functions. One that checks if an expression is a positive int constant `&Expr -> Option<u128>`, and another that takes two expressions and returns one as a constant `(&Expr, &Expr) -> Option<(u128, &Expr)>`). This will remove the need for the bool argument and get rid of the extra check for `ExprKind::Binary`